server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -240,24 +258,29 @@ interface CacheEntity extends Accountable {
          * Called when this entity instance is removed
          */
         void onRemoval(RemovalNotification<Key, BytesReference> notification);
+
     }
 
     /**
      * Unique key for the cache
      *
      * @opensearch.internal
      */
-    static class Key implements Accountable {
-        private static final long BASE_RAM_BYTES_USED = RamUsageEstimator.shallowSizeOfInstance(Key.class);
-
+    class Key implements Accountable, Writeable {
         public final CacheEntity entity; // use as identity equality
```

Comment:
Given the comment here and the definition of `equals` (which right now is looking for equal `IndexShard` references), I feel like you're better off moving away from the existing `CacheEntity` idea altogether.

Right now, it's relying on a cache key being associated with a given instance of `IndexShard`, which is no longer the case now that you're planning on serializing the cache keys. If that's the case, why not just use `ShardId` (instead of `IndexShard`) to determine equality? You could avoid needing to pass `IndicesService` around.

Reply:
It does make sense to use ShardId.

But I can't remove IndexShard as a cacheEntity as it is used to collect stats per indexShard.
For instance look out for code lines like ```key.entity.onMiss();``` or ```key.entity.onHit();```. 
All these cache stats are collected at indexShard level, thats why can't move away from IndexShard. Just using shardId wouldn't allow me to do that, until unless there we extract IndexShard from ShardId which is what we are doing anyway.

Reply:
I would suggest a bigger cleanup, honestly. This whole thing is a huge mess. 

Why should a component of a cache key be responsible for tracking statistics about cache access related to that component of the key?

I think we could detangle this mess with the following changes:
1. Change the return type of `getCacheIdentity` to `Writeable` (i.e. it must be serializable).
2. Remove `CacheEntity` from `Key`, replacing it with `Writeable cacheIdentity`. Instead of holding a reference to the whole `CacheEntity`, it holds a reference to its identity.
3. The `getOrCompute` method in `IndicesRequestCache` should use the `cacheEntity` parameter instead of `key.entity` for the `onHit()` and `onMiss()` calls. It should also still pass `cacheEntity` to the `CleanupKey` constructor.
4. The `cleanCache` method almost works, except for the following part:
```
if (currentKeysToClean.contains(new CleanupKey(key.entity, key.readerCacheKey))) {
    iterator.remove();
}
```

IMO, the best solution there is to change the type of `currentKeysToClean` from `Set<CleanupKey>` to `Set<Tuple<Writeable, String>>` (i.e. the elements are cache entity identities and reader cache key IDs).

Reply:
>Why should a component of a cache key be responsible for tracking statistics about cache access related to that component of the key?

It is mostly related to handle all those callbacks like ```onRemoval``` from cache where you have access to only key and value for entry. As cacheEntity is present as part of key, we are easily able to collect stats per shard. Without which it is hard/impossible.

Also in our new [framework](https://github.com/opensearch-project/OpenSearch/pull/10753/files#diff-3e24faf09cb1541791033327944a9eb18442a2350909b36f870e86623b68c4b1) changes, we are also handling hits/misses as part of listener/callbacks similar to ```onRemoval``` above(see StoreAwareCacheEventListener). In this case we ideally will have access to only key/value for an entry and not cacheEntity.

I think it's better to keep the key structure as it is and not change it fundamentally. Personally I don't think current changes are a mess as such.

Reply:
This was changed as per discussion. Using ShardId as an entity within Key object.